// @nebular/theme/components/tree-grid/data-source/tree-grid.model.js
/*
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
export var DEFAULT_ROW_LEVEL = 0;
/**
 * Implicit context of cells and rows
 */
var NbTreeGridPresentationNode = /** @class */ (function () {
    function NbTreeGridPresentationNode(node, level) {
        if (level === void 0) { level = DEFAULT_ROW_LEVEL; }
        this.node = node;
        this.level = level;
        this.children = [];
    }
    Object.defineProperty(NbTreeGridPresentationNode.prototype, "expanded", {
        /**
         * Row expand state
         */
        get: function () {
            return this.node.expanded;
        },
        set: function (value) {
            this.node.expanded = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NbTreeGridPresentationNode.prototype, "data", {
        /**
         * Data object associated with row
         */
        get: function () {
            return this.node.data;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * True if row has child rows
     */
    NbTreeGridPresentationNode.prototype.hasChildren = function () {
        return !!this.children && !!this.children.length;
    };
    return NbTreeGridPresentationNode;
}());
export { NbTreeGridPresentationNode };
